### ISSUE TITLE
Computer freezes after 1 minute

### ISSUE THREAD

**Issue by davy7125** [torculus/DownFall#1]
After installing the extension a lot of "ABC" are falling - this is the expected behavior - but then everything is slowed down and then the computer freezes. I'm running Gnome 40 on Fedora 34, fresh install.

**Comment by dupr0n** [comment]
@torculus  I'm kind of a newbie, but I'm trying to understand this. From what I understand, you have a "MAX_ITEM"s number of "FallItem"s and you're calling the "fall" method for each, which then animates and then gets hidden. Once everything is done falling, then what happens? Is the "dropItems" method called again? If so, shouldn't each "FallItem" be deleted instead of hidden?

Also, could you suggest how to get this setup and running locally in my own machine so I could test it?

**Comment by torculus** [comment]
@dupr0n, thanks for wanting to help! So from a top-down perspective, here is what happens when the extension is activated:

1. A FallItemsManager (FIM) is created
2. The FIM creates itemContainer (a [Clutter.Actor](https://gjs-docs.gnome.org/clutter7~7_api/clutter.actor)) and a matrixItemsContainer; then adds them with `Main.uiGroup.add_actor()`
3. The FIM then creates MAX_ITEMS FallItems, adds them to itemContainer (for easy removal on `disable()`), and calls `fall()` on each of them (the `//make it rain` part)
4. `fall()` uses Clutter to move the item across the screen. When the transition is completed, the FallItem calls `this.finish()` which
    1. calls `hide()` on the FallItem
    2. creates the Firework burst (if set)
    3. resets the position on FallItem and calls `fall()` again after a delay with `GLib.idle_add()`
5. Step 4 continues as long as the extension is running
6. When the extension is disabled, the itemContainer, matContainer, and FIM are all destroyed

I finally added the extensions.gnome.org link [here](https://extensions.gnome.org/extension/3539/downfall/), so you can test this locally by installing it from there (just don't run it for more than 30 seconds, to be safe). GNOME extensions get installed to `home/<YOUR_USERNAME>/.local/share/gnome-shell/extensions/`, so that's where you'll find the code and be able to modify it

Since you say you're kind of a newbie, here are some helpful tips:

1. GNOME shell extensions are written in "gnome javascript", so you'll want to search "*gjs*" documentation (most GNOME documentation is in C)
2. Pressing `Alt + F2` and then `lg` will bring up GNOME's "looking glass" debugger, which allows you to navigate directly to the code
3. Running `joournalctl /usr/bin/gnome-shell` in a terminal will let you see real-time errors with GNOME shell (but it doesn't help if your computer is frozen)

Happy hacking!

**Comment by torculus** [comment]
Guys, I fixed it!

Turns out that the `startEndPoints()` function in `utils.js` was doing *WAY* too much algebra and could be simplified with [this](http://webreflection.github.io/gjs-documentation/GLib-2.0/GLib.random_int_range.html), leading to the lockup

Just to be sure, I tested it for about 5 minutes. It works! I'll be releasing version 3.0.0 shortly

If you encounter any other freezes, please let me know or open a new issue!

**Comment by dupr0n** [comment]
> Guys, I fixed it!
> 
> Turns out that the `startEndPoints()` function in `utils.js` was doing _WAY_ too much algebra (leading to the lockup) and could be simplified with [this](http://webreflection.github.io/gjs-documentation/GLib-2.0/GLib.random_int_range.html)
> 
> Just to be sure, I tested it for about 5 minutes. It works! I'll be releasing version 3.0.0 shortly
> 
> If you encounter any other freezes, please let me know or open a new issue!

Great! I'll give that version a try once it's out!
